### PR TITLE
ADC - AD7192 library

### PR DESCRIPTION
**Pull request by artyLee**
### Problem

Implement a 24 bit ADC - AD7192 SPI library.

### Solution

Implement some driver functions as below:
```
void ad7192Init(void);
void ad7192SoftwareReset(void);
void ad7192SetPGAGain(uint8_t gain);
void ad7192SetChannel(uint8_t channel);
void ad7192SetFilterSelectBit(uint16_t filterRate);
void ad7192InternalZeroScaleCalibration(void);
void ad7192InternalFullScaleCalibration(void);
void ad7192InternalZeroFullScaleCalibration(void);
void ad7192StartSingleConversion(void);
void ad7192StartContinuousConversion(void);
void ad7192WriteRegisterValue(uint8_t registerAddress, uint32_t registerValue, uint8_t bytesSize);
uint32_t ad7192ReadRegisterValue(uint8_t registerAddress, uint8_t bytesSize);
uint32_t ad7192ReadConvertingData(void);
uint32_t ad7192ReadADCChannelData(uint8_t channel);
uint8_t ad7192ReadIDInfor(void);
void ad7192ReadRegisterMapValue(void);
float ad7192RawDataToVoltage(uint32_t rawData);
float ad7192TempSensorDataToDegC(uint32_t rawData);
```

### Steps to Test

The library was compiled by the Particle deviceOS based on the v2.x version. It is working fine on the Argon board.

### Example App

```c

#include "ad7192.h"
#include "ad7192_defs.h"
#include "application.h"

Serial1LogHandler logHandler(115200, LOG_LEVEL_WARN, {{"app",    LOG_LEVEL_ALL}});
SYSTEM_MODE(MANUAL);
void setup() {
    ad7192Init();
    ad7192SetPGAGain(1);
    ad7192SetFilterSelectBit(100);
    ad7192InternalZeroFullScaleCalibration();
}

void loop() {
    uint32_t rawData = ad7192ReadADCChannelData(AD7192_CH_AIN1);
    float volt = ad7192RawDataToVoltage(rawData);
    Log.info("CH1 data: %lx, Voltage Measurement: %fmV", rawData, volt);
    delay(1000);
}
```

### References

- AD7192 Datasheet
- SPI communication

---

### Completeness

- [ ] Run unit/integration/application tests on device
- [ ] Added documentation

### CODE REVIEW

File: AD7192/ad7192_defs.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+/* AD7192 Register Map */
+#define AD7192_REG_COMM         0 // Communications Register (WO, 8-bit)
```
Comment (YutingYou) [inline AD7192/ad7192_defs.h]: Nit: please remove the extra space, check other files too.
image

**Comment by YutingYou** [comment]
File structure could be:
AD7192
- docs
- src
- examples

File: AD7192/ad7192.cpp [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        byteIndex++;
+    }
+    digitalWrite(CS_PIN, HIGH);
+    SPI.endTransaction();
+
+    return(readRegisterValue);
+}
+
+void ad7192WriteRegisterValue(uint8_t registerAddress, uint32_t registerValue, uint8_t bytesSize) {
+    uint8_t txBuffer[4] = {0, 0, 0, 0};
```
Comment (YutingYou) [inline AD7192/ad7192.cpp]: `uint8_t txBuffer[4] = {0}`

File: AD7192/ad7192.cpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+void ad7192SoftwareReset(void) {
+    SPI.beginTransaction(SPISettings(4*MHZ, MSBFIRST, SPI_MODE3));
+    digitalWrite(CS_PIN, LOW);
+    for(char i = 0; i < 6; i++) { // If a Logic 1 is written to the AD7192 DIN line for at least 40 serial clock cycles, the serial interface is reset
+        SPI.transfer(0xFF);
+    }
+    digitalWrite(CS_PIN, HIGH);
+    SPI.endTransaction();
+    delay(1);                   // Following a reset, the user should wait a period of 500 μs before addressing the serial interface
```
Comment (YutingYou) [inline AD7192/ad7192.cpp]: 👍 for the comment, every delay should be meaningful

File: AD7192/ad7192.cpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    ad7192WriteRegisterValue(regAddress, registerMap[regAddress], registerSize[regAddress]); // After overwriting setting value to mode register
+}
+
+void ad7192InternalZeroFullScaleCalibration(void) {
+    ad7192InternalZeroScaleCalibration();
+    ad7192InternalFullScaleCalibration();
+}
+
+void ad7192SetFilterSelectBit(uint16_t filterRate) {
+    if(filterRate > 0x3ff) {
```
Comment (YutingYou) [inline AD7192/ad7192.cpp]: What is 0x3FF? Would be better to use a macro definition or leave a comment.

File: AD7192/ad7192.cpp [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    SPI.transfer(writeCommandByte);
+    while(byteIndex < bytesSize) {
+        receiveBuffer = SPI.transfer(0);
+        readRegisterValue = (readRegisterValue << 8) + receiveBuffer;
+        byteIndex++;
+    }
+    digitalWrite(CS_PIN, HIGH);
+    SPI.endTransaction();
+
+    return(readRegisterValue);
```
Comment (YutingYou) [inline AD7192/ad7192.cpp]: Weird, `return readRegisterValue;`, please also correct other lines.

File: AD7192/ad7192.cpp [hunk 6]
```diff
@@ -0,0 +1,217 @@
+#include "ad7192.h"
+#include "ad7192_defs.h"
+
+#define CS_PIN                  D5 // Chip select pin
+#define V_REF                   3300 // millivolt reference voltage
```
Comment (YutingYou) [inline AD7192/ad7192.cpp]: Nit: add a new line here

File: AD7192/ad7192.h [hunk 7]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+#include "application.h"
```
Comment (YutingYou) [inline AD7192/ad7192.h]: Please include the head files as less as possible. Move it to the cpp file.

File: AD7192/examples/ad7192singlechannelconversion.cpp [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ * Project AD7192
+ * Description:
+ * Author:
+ * Date:
+ */
+#include "ad7192.h"
+#include "ad7192_defs.h"
+#include "application.h"
+
+Serial1LogHandler logHandler(115200, LOG_LEVEL_WARN, {{"app",    LOG_LEVEL_ALL}});
```
Comment (YutingYou) [inline AD7192/examples/ad7192singlechannelconversion.cpp]: Nit: newline between different code block

File: AD7192/userapp.cpp [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+ */
+#include "ad7192.h"
+#include "ad7192_defs.h"
+#include "application.h"
+
+
+Serial1LogHandler logHandler(115200, LOG_LEVEL_WARN, {{"app",    LOG_LEVEL_ALL}});
+SYSTEM_MODE(MANUAL);
+void setup() {
+
```
Comment (YutingYou) [inline AD7192/userapp.cpp]: Nit: please rework all the files to remove extra newlines and add newline between different code block.

**Comment by artyLee** [comment]
> File structure could be:
> AD7192
> 
> * docs
> * src
> * examples

Has been fixed. But looks my local machine not sensitive file naming case.

**Review by YutingYou — APPROVED** [review]

File: AD7192/ad7192_defs.h (hunk 1 above)
Comment (artyLee) [reply AD7192/ad7192_defs.h]: This is a latent detail. Thank you figure out.

File: AD7192/ad7192.cpp (hunk 2 above)
Comment (artyLee) [reply AD7192/ad7192.cpp]: Optimize option

File: AD7192/ad7192.cpp (hunk 4 above)
Comment (artyLee) [reply AD7192/ad7192.cpp]: Yes, the setting number here is not clear for reading.

File: AD7192/ad7192.cpp (hunk 5 above)
Comment (artyLee) [reply AD7192/ad7192.cpp]: You mean when using `return` keyword should not include `() `, so correct format should be`return (value);`changed to `return value;`

File: AD7192/ad7192.cpp (hunk 6 above)
Comment (artyLee) [reply AD7192/ad7192.cpp]: Sounds good.

File: AD7192/ad7192.h (hunk 7 above)
Comment (artyLee) [reply AD7192/ad7192.h]: Yes, I think so. The reason for this `#include "application.h"` is when I compile it doesn't found some data type, such as:`uint8_t, uint32_t`.

File: AD7192/examples/ad7192singlechannelconversion.cpp (hunk 8 above)
Comment (artyLee) [reply AD7192/examples/ad7192singlechannelconversion.cpp]: OK, thanks

File: AD7192/userapp.cpp (hunk 9 above)
Comment (artyLee) [reply AD7192/userapp.cpp]: Aha, ok.

**Pull request merged**